### DISCUSSION TITLE
Editor Test Scene throws an error for custom rulesets

### DISCUSSION THREAD

**Discussion by naoei** [ppy/osu#16668] [Q&A] [answered]
While updating my custom ruleset to the new update to make it compatible, I've noticed that one of the test scenes always throws an error, and it couldn't be fixed on my end. [This test scene](https://github.com/taulazer/tau/blob/master/osu.Game.Rulesets.Tau.Tests/TestSceneTauEditor.cs) gave me this error:
```cs
[runtime] 2022-01-27 23:29:51 [error]: System.ArgumentException: Only beatmaps in the osu, taiko, catch, or mania rulesets can be encoded to the legacy beatmap format. (Parameter 'beatmap')
[runtime] 2022-01-27 23:29:51 [error]: at osu.Game.Beatmaps.Formats.LegacyBeatmapEncoder..ctor(IBeatmap beatmap, ISkin skin)
[runtime] 2022-01-27 23:29:51 [error]: at osu.Game.Screens.Edit.EditorChangeHandler.UpdateState()
[runtime] 2022-01-27 23:29:51 [error]: at osu.Game.Screens.Edit.TransactionalCommitComponent.SaveState()
[runtime] 2022-01-27 23:29:51 [error]: at osu.Game.Screens.Edit.EditorChangeHandler..ctor(EditorBeatmap editorBeatmap)
[runtime] 2022-01-27 23:29:51 [error]: at osu.Game.Screens.Edit.Editor.load(OsuColour colours, OsuConfigManager config)
```

After looking through the codebase a little bit, I've spotted that this line gets run during the test scene.
https://github.com/ppy/osu/blob/d1158acb828e0e0b5c4fe97621f0c5f3b7010bf6/osu.Game/Screens/Edit/EditorChangeHandler.cs#L62-L63

**Comment by andy840119** [comment]
https://github.com/karaoke-dev/karaoke/blob/master/osu.Game.Rulesets.Karaoke.Tests/Beatmaps/TestKaraokeBeatmap.cs#L16
This is an old issue that happened for a while.
Before waiting for lazer to support customized ruleset in the editor change handler, you can use this tricky way to avoid that.

**Reply by naoei** [reply]
I don't mind waiting for proper custom ruleset encoding to be supported. However, I have no portion of code that touches the encoder or anything of the like, so in my expectations, this shouldn't break. Besides, this only broke very recently with the proper migration of realms.

**Reply by andy840119** [reply]
Because there's no other encoder for custom ruleset, so it can expected that it will trow exception from the check.
IDK it will be fixed by official(maybe add an empty encoder for custom ruleset for now?) or maybe will be waited until custom encoder implemented.

**Comment by peppy** [comment, marked as answer]
This seems to imply that your ruleset is current when that test runs. Can you give some more details on how you're running the tests? Also can you attach your `client.realm` for reference?

**Reply by naoei** [reply]
I'm not entirely sure what you mean by giving more details on how I'm running the tests? Everything is up to date (2022.128.0) and I'm running on Release mode if that helps.

Either way, here's the [client.realm](https://github.com/ppy/osu/files/7955419/client.zip) file for reference.

**Reply by peppy** [reply]
Ah I see now that it's a custom test you have setup (thought you were referring to running `osu.Game.Tests`). I guess this would mean the editor also crashes when you enter it via non-tests as well? I'd presume that `EditorChangeHandler` is fired in both cases.

**Reply by naoei** [reply]
Yep!

**Reply by peppy** [reply]
Thanks. Given, you're not going to be able to save out to a file at the end of the day due to the reason mentioned in that exception, but I guess you still want to be able to enter / work on the editor regardless? Would it be an improvement if the exception only fired when you tried to explicitly save?

**Reply by naoei** [reply]
Definitely! I believe that's how it was before right? I was moreover waiting until a proper method of serialization/deserialization comes up for custom rulesets.

**Reply by peppy** [reply]
I think previously it was able to start due to a bug in implementation that was fixed with recent changes.

**Reply by peppy** [reply]
Should be resolved via #16676.
